sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +163,169 @@ class CollationStringExpressionsSuite
+    }
+  }
+
+  test("left/right/substr on explicitly collated proper string returns proper value and type") {
+    case class QTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        QTestCase("select left('abc' collate " + c + ", 1)", c, Row("a")),
+        QTestCase("select right('def' collate " + c + ", 1)", c, Row("f")),
+        QTestCase("select substr('abc' collate " + c + ", 2)", c, Row("bc")),
+        QTestCase("select substr('example' collate " + c + ", 0, 2)", c, Row("ex")),
+        QTestCase("select substr('example' collate " + c + ", 1, 2)", c, Row("ex")),
+        QTestCase("select substr('example' collate " + c + ", 0, 7)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 1, 7)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 0, 100)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 1, 100)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 2, 2)", c, Row("xa")),
+        QTestCase("select substr('example' collate " + c + ", 1, 6)", c, Row("exampl")),
+        QTestCase("select substr('example' collate " + c + ", 2, 100)", c, Row("xample")),
+        QTestCase("select substr('example' collate " + c + ", 0, 0)", c, Row("")),
+        QTestCase("select substr('example' collate " + c + ", 100, 4)", c, Row("")),
+        QTestCase("select substr('example' collate " + c + ", 0, 100)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 1, 100)", c, Row("example")),
+        QTestCase("select substr('example' collate " + c + ", 2, 100)", c, Row("xample")),
+        QTestCase("select substr('example' collate " + c + ", -3, 2)", c, Row("pl")),
+        QTestCase("select substr('example' collate " + c + ", -100, 4)", c, Row("")),
+        QTestCase("select substr('example' collate " + c + ", -2147483648, 6)", c, Row("")),
+        QTestCase("select substr(' a世a ' collate " + c + ", 2, 3)", c, Row("a世a")), // scalastyle:ignore
+        QTestCase("select left(' a世a ' collate " + c + ", 3)", c, Row(" a世")), // scalastyle:ignore
+        QTestCase("select right(' a世a ' collate " + c + ", 3)", c, Row("世a ")), // scalastyle:ignore
+        QTestCase("select substr('AaAaAaAa000000' collate " + c + ", 2, 3)", c, Row("aAa")),
+        QTestCase("select left('AaAaAaAa000000' collate " + c + ", 3)", c, Row("AaA")),
+        QTestCase("select right('AaAaAaAa000000' collate " + c + ", 3)", c, Row("000")),
+        QTestCase("select substr('' collate " + c + ", 1, 1)", c, Row("")),
+        QTestCase("select left('' collate " + c + ", 1)", c, Row("")),
+        QTestCase("select right('' collate " + c + ", 1)", c, Row("")),
+        QTestCase("select left('ghi' collate " + c + ", 1)", c, Row("g"))
```

Comment:
I don't think we need this many test cases here, if you didn't modify the way Substring/Left/Right expressions behave when given collated strings (i.e. you didn't introduce any collation awareness to nullSafeEval/doCodeGen), then there should be no need to go this deep - a couple of test cases should do the trick just fine